framework/doc/content/source/runners/Runner.md

```diff
@@ -0,0 +1,184 @@
+modified from within the input file via a runner's `begin_exec_flag` and
+`end_exec_flag` input parameters.
+
+## Writing Custom Runners
+
+Runners have one primary function - `virtual Result gogogadget()` - that must
+be implemented.  If a runner has any internal runners, it will call these
+runners' `Result run()` functions - NOT their gogogadget functions.
+
+All runners' "gogogadget" and "run" functions return a `Result` value contains
+information about how execution turned out within the runner tree.  Each
+runner is responsible for recording how convergence/success occurs within it.
+This should generally be accomplished using the `Result::pass(msg)` and
+`Result::fail(msg)` functions on a result object created and initialized
+within it's own gogogadget function:
+
+```
+Result
+FooRunner::gogogadget()
+{
+  Result r(this);
+
+  ...
+  bool success = ... // do some solve stuff
+
+  if (!success)
+    r.fail("the foo didn't work right with the bar");
+  else
+  {
+    // by default, a result is considered successful/converged - so we only need
+    // to call fail on failure - and calling pass on success is optional.
+    r.pass("runnin' like a well oiled machine");
+  }
+
+  ...
+
+  return r;
+}
+```
+
+Some runners will have internal/sub runners that they need to execute.  They
+are both responsible for initiating this execution as well as recording the
+result value generated by these runners using the `Result::record` function:
+
+```
+
+InputParameters
+Steady2::validParams()
+{
+  InputParameters params = Runner::validParams();
+  // create input parameters for our sub/internal solve runners
+  params.addRequiredParam<std::string>("solve1", "the first solve");
+  params.addRequiredParam<std::string>("solve2", "the second solve");
+  return params;
+}
+
+FooRunner::FooRunner(InputParameters & params)
+  : _inner_solve1(&_fe_problem.getRunner(getParam<std::string>("solve1"))), // retrieve inner runner objects
+    _inner_solve2(&_fe_problem.getRunner(getParam<std::string>("solve2")))
+{
+}
+
+Result
+FooRunner::gogogadget()
+{
+  Result r(this);
+  ...
+  // When we record an inner/sub runner's result, we give it a label - which
+  // helps identify its placement/role within the runner heirarchy.
+  r.record("solve1", _inner_solve1->run());
+  r.record("solve2", _inner_solve2->run());
+  ...
+  return r;
+}
+```
+
+Result values provide a convenience `bool convergedAll()`function for
+recursively determining if *any* single runner result within the currently
+executed portion of the tree has failed to converge.  When checking for convergence within a runner, this is usually the mechanism that should be used:
+
+```
+Result
+FooRunner::gogogadget()
+{
+  Result r(this);
+  ...
+  // When we record an inner/sub runner's result, we give it a label - which
+  // helps identify its placement/role within the runner heirarchy.
+  r.record("solve1", _inner_solve1->run());
+  r.record("solve2", _inner_solve2->run());
```

Comment:
Since we have the outer wrapping `run()` function... can we just use that to do the recording instead?  This seems like it may be error-prone.

Reply:
Record is for building the result hierarchy.  And the parent/base class doesn't know anything about the child runners and details of particular runner implementations.  A runner calls its child/sub runners at its leisure - not at particular points we can know a prior.  As it goes along and results are generated they are accumulated into a result tree.  The only way to avoid this is to develop a shadow-hierarchy based on input file syntax that constructs the result tree to mirror runner structure - which seems more restrictive and a lot more complicated for what I would consider relatively small benefit.

Reply:
Hmmm... I can see what you're saying... let me think about it a bit more.